### PR TITLE
Describe status codes on the EventEngine API

### PR DESCRIPTION
**Pull request by drfloob**
@markdroth 
@nicolasnoble

This continues the work of ironing out EventEngine API details.

Note also that the `AcceptCallback` has lost its status parameter. In short: that callback will never be called in non-OK situations - either CreateListener will fail, or its `on_shutdown` callback will be called with an error status.

Some of these changes also appear in PR https://github.com/grpc/grpc/pull/26026. For example, the addition of default destructors.

### CODE REVIEW

File: include/grpc/event_engine/event_engine.h [hunk 1]
```diff
@@ -144,40 +154,67 @@ class EventEngine {
     virtual const ResolvedAddress* GetLocalAddress() const = 0;
   };
 
-  /// Called when a new connection is established. This callback takes ownership
-  /// of the Endpoint and is responsible for its destruction.
-  using OnConnectCallback = std::function<void(absl::Status, Endpoint*)>;
+  /// Called when a new connection is established.
+  ///
+  /// If the connection attempt was not successful, implementations should pass
+  /// the appropriate statuses to \a on_read. For example, callbacks might
```
Comment (markdroth) [inline include/grpc/event_engine/event_engine.h:160]: s/to \a on_read/to this callback/

File: include/grpc/event_engine/event_engine.h [hunk 2]
```diff
@@ -144,40 +154,67 @@ class EventEngine {
...
+  /// callback will only be called in healthy scenarios where connections can be
+  /// accepted.
+  ///
+  /// \a on_shutdown will be called exactly once, with either an OK status for a
+  /// normal \a Listener or \a EventEngine shutdown, or a non-OK status if the
+  /// Listener or gRPC encountered an error.
+  ///
+  /// This method may return a non-OK status immediately if an error was
+  /// encountered in any synchronous steps required to create a \a Listener. If
+  /// this happens, neither \a on_shutdown nor \a on_accept callbacks will have
```
Comment (markdroth) [inline include/grpc/event_engine/event_engine.h:196]: This seems a little confusing, in light of the statement in the previous paragraph that on_shutdown will be called exactly once.  I suggest rewriting these two paragraphs as follows:

"""
This method may return a non-OK status immediately if an error was encountered in any synchronous steps required to create the Listener.  In this case, \a on_shutdown will never be called.

If this method returns a Listener, then \a on_shutdown will be invoked exactly once, when the Listener is shut down.  The status passed to it will indicate if there was a problem during shutdown.
"""

I don't think it's necessary to say that on_accept won't be called if there is no Listener, because if there is no Listener, then there are by definition no connections being accepted.

Also, I didn't say that the Listener will be shut down when the EventEngine is shut down, because I thought we said that EventEngine shutdown would wait for all current activity to complete, and the Listener is one such activity.  In other words, I think EventEngine shutdown would not complete until the Listener is shut down.

File: include/grpc/event_engine/event_engine.h [hunk 3]
```diff
@@ -208,42 +245,59 @@ class EventEngine {
...
-    // TODO(hork): define status codes for the callback
-    /// Asynchronously resolve an address. \a default_port may be a non-numeric
-    /// named service port, and will only be used if \a address does not already
-    /// contain a port component.
+    /// Asynchronously resolve an address.
+    ///
+    /// \a default_port may be a non-numeric named service port, and will only
+    /// be used if \a address does not already contain a port component.
+    ///
+    /// \a on_resolve will be called exactly once. Implementations are expected
```
Comment (markdroth) [inline include/grpc/event_engine/event_engine.h:255]: Suggest wording this similarly to `Endpoint::Read()` above:

"""
The \a on_resolve callback will be called when the lookup is complete.  Implementations should pass the appropriate statuses to the callback. For example, callbacks might expect to receive DEADLINE_EXCEEDED when the deadline is exceeded or CANCELLED if the lookup was cancelled.
"""

**Review by markdroth — COMMENTED** [review]
Only a few minor comments; overall, this looks good!

File: include/grpc/event_engine/event_engine.h (hunk 1 above)
Comment (drfloob) [reply include/grpc/event_engine/event_engine.h]: Done.

File: include/grpc/event_engine/event_engine.h (hunk 2 above)
Comment (drfloob) [reply include/grpc/event_engine/event_engine.h]: Thanks for the improvement and the correction. You're right, we had decided that the EventEngine's shutdown will wait for the Listener to complete its business - a soft shutdown operation.

File: include/grpc/event_engine/event_engine.h (hunk 3 above)
Comment (drfloob) [reply include/grpc/event_engine/event_engine.h:254]: I avoided saying "when the lookup is complete" because even though it's true (e.g., a lookup failure counts as a completion), that wording on its own might be interpreted as "when the lookup has completed successfully" which is only part of the story.

How about changing the first sentence to:

"""
The \a on_resolve callback will be called when the lookup has completed successfully or failed.
"""

Comment (markdroth) [reply include/grpc/event_engine/event_engine.h:254]: I don't actually think that "when the lookup is complete" implies anything about the success or failure of the lookup, especially since the rest of the paragraph talks about the status being passed to the callback.  But if you're concerned about that, how about this wording:

"""
When the lookup is complete, the \a on_resolve callback will be invoked with a status indicating the success or failure of the lookup.  Implementations should pass the appropriate statuses to the callback. For example, callbacks might expect to receive DEADLINE_EXCEEDED when the deadline is exceeded or CANCELLED if the lookup was cancelled.
"""

Hmmm... Would it make sense to change the callback arguments from `absl::Status` and `std::vector<ResolvedAddress>` to a single argument of type `absl::StatusOr<std::vector<ResolvedAddress>>`?   That way, the API makes it clearer that it will either fail or get a result -- i.e., there will never be a case where both arguments are relevant at the same time.

Comment (drfloob) [reply include/grpc/event_engine/event_engine.h]: That language works, thanks. And yes, StatusOr would make sense for the Lookup callbacks and OnConnectCallback. I've made those changes.

**Review by markdroth — APPROVED** [review]
Looks great!
